README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# zscorer: Anthropometric Z-Score Calculator <img src="man/figures/logo.png" width="200" align="right" />

<!-- badges: start -->

[![Project Status: Active – The project has reached a stable, usable
state and is being actively
developed.](http://www.repostatus.org/badges/latest/active.svg)](http://www.repostatus.org/#active)
[![lifecycle](https://img.shields.io/badge/lifecycle-maturing-blue.svg)](https://www.tidyverse.org/lifecycle/#maturing)
[![CRAN\_Status\_Badge](http://www.r-pkg.org/badges/version/zscorer)](https://cran.r-project.org/package=zscorer)
[![CRAN](https://img.shields.io/cran/l/zscorer.svg)](https://CRAN.R-project.org/package=zscorer)
[![CRAN](http://cranlogs.r-pkg.org/badges/zscorer)](https://CRAN.R-project.org/package=zscorer)
[![CRAN](http://cranlogs.r-pkg.org/badges/grand-total/zscorer)](https://CRAN.R-project.org/package=zscorer)
[![R-CMD-check](https://github.com/nutriverse/zscorer/workflows/R-CMD-check/badge.svg)](https://github.com/nutriverse/zscorer/actions)
[![AppVeyor Build
Status](https://ci.appveyor.com/api/projects/status/github/nutriverse/zscorer?branch=master&svg=true)](https://ci.appveyor.com/project/nutriverse/zscorer)
[![test-coverage](https://github.com/nutriverse/zscorer/workflows/test-coverage/badge.svg)](https://github.com/nutriverse/zscorer/actions)
[![Codecov test
coverage](https://codecov.io/gh/nutriverse/zscorer/branch/master/graph/badge.svg)](https://codecov.io/gh/nutriverse/zscorer?branch=master)
[![DOI](https://zenodo.org/badge/119683584.svg)](https://zenodo.org/badge/latestdoi/119683584)
[![CodeFactor](https://www.codefactor.io/repository/github/nutriverse/zscorer/badge)](https://www.codefactor.io/repository/github/nutriverse/zscorer)
<!-- badges: end -->

Nutritional status, especially in children, has been widely assessed
using anthropometric measures in both developing and developed
countries. Once a childs’ anthropometry has been correctly measured and
their age has been recorded, their growth and nutritional status can be
assessed using a standardised age- and sex-specific growth reference to
calculate z-scores. These z-scores express anthropometric measurements
as several standard deviations (SDs) below or above the reference mean
or median value. The calculation of anthropometric z-scores is one of
the most widely-used analytical technique used in nutrition surveys and
nutrition research.

## What does `zscorer` do?

The `zscorer` packages facilitates the calculation of a range of
anthropometric *z-scores* (i.e. the number of standard deviations from
the mean) and adds them to survey data:

  - **Weight-for-length (wfl)** z-scores for children with lengths
    between 45 and 110 cm

  - **Weight-for-height (wfh)** z-scores for children with heights
    between 65 and 120 cm

  - **Length-for-age (lfa)** z-scores for children aged less than 24
    months

  - **Height-for-age (hfa)** z-scores for children aged between 24 and
    228 months

  - **Weight-for-age (wfa)** z-scores for children aged between zero and
    120 months

  - **Body mass index-for-age (bfa)** z-scores for children aged between
    zero and 228 months

  - **MUAC-for-age (mfa)** z-scores for children aged between 3 and 228
    months

  - **Triceps skinfold-for-age (tsa)** z-scores for children aged
    between 3 and 60 months

  - **Sub-scapular skinfold-for-age (ssa)** z-scores for children aged
    between 3 and 60 months

  - **Head circumference-for-age (hca)** z-scores for children aged
    between zero and 60 months

The *z-scores* are calculated using the **WHO Child Growth
Standards**<sup>1,2</sup> for children aged between zero and 60 months
or the **WHO Growth References**<sup>3</sup> for school-aged children
and adolescents. MUAC-for-age (mfa) z-scores for children aged between
60 and 228 months are calculated using the MUAC-for-age growth reference
developed by Mramba et al. (2017)<sup>4</sup> using data from the USA
and Africa. This reference has been validated with African school-age
children and adolescents. The `zscorer` comes packaged with the WHO
Growth References data and the MUAC-for-age reference data.

## Installation

You can install `zscorer` from [CRAN](https://cran.r-project.org):

``` r
install.packages("zscorer")
```

or you can install the development version of `zscorer` from
[GitHub](https://github.com/nutriverse/zscorer) with:

``` r
if(!require(remotes)) install.packages("remotes")
remotes::install_github("nutriverse/zscorer")
```

then load `zscorer`

``` r
# load package
library(zscorer)
```

## Usage - WHO Growth Standards functions

### Calculating anthropometric z-scores using the addWGSR() function

The main function in the `zscorer` package is `addWGSR`.

To demonstrate its usage, we will use the accompanying dataset in
`zscorer` called `anthro3`. We inspect the dataset as follows:

``` r
head(anthro3)
```

which returns:

    #>   psu age sex weight height muac oedema
    #> 1   1  10   1    5.7   64.2  125      2
    #> 2   1  10   2    5.8   64.4  121      2
    #> 3   1   9   2    6.5   62.2  139      2
    #> 4   1  11   9    6.5   64.9  129      2
    #> 5   1  24   2    6.5   72.9  120      2
    #> 6   1  12   2    6.6   69.4  126      2

`anthro3` contains anthropometric data from a Rapid Assessment Method
(RAM) survey from Burundi.

Anthropometric indices (e.g. weight-for-height z-scores) have not been
calculated and added to the data.

We will use the `addWGSR` function to add weight-for-height (wfh)
z-scores to the example data:

``` r
svy <- addWGSR(data = anthro3, sex = "sex", firstPart = "weight",
               secondPart = "height", index = "wfh")
#> ================================================================================
```

A new column named **wfhz** has been added to the dataset:

    #>   psu age sex weight height muac oedema  wfhz
    #> 1   1  10   1    5.7   64.2  125      2 -2.73
    #> 2   1  10   2    5.8   64.4  121      2 -2.04
    #> 3   1   9   2    6.5   62.2  139      2  0.13
    #> 4   1  11   9    6.5   64.9  129      2    NA
    #> 5   1  24   2    6.5   72.9  120      2 -3.44
    #> 6   1  12   2    6.6   69.4  126      2 -2.26

The `wfhz` column contains the weight-for-height (wfh) z-scores
calculated from the `sex`, `weight`, and `height` columns in the
`anthro3` dataset. The calculated z-scores are rounded to two decimals
places unless the `digits` option is used to specify a different
precision (run `?addWGSR` to see description of various parameters that
can be specified in the `addWGSR` function).

The `addWGSR` function takes up to nine parameters to calculate each
index separately, depending on the index required. These are described
in the *Help* files of the `zscorer` package which can be accessed as
follows:

``` r
?addWGSR
```

The **standing** parameter specifies how “stature” (i.e. length or
height) was measured. If this is not specified, and in some special
circumstances, height and age rules will be applied when calculating
z-scores. These rules are described in the table below.

 

| **index**  | **standing** | **age**     | **height**       | **Action**                           |
| ---------- | ------------ | ----------- | ---------------- | ------------------------------------ |
| hfa or lfa | standing     | \< 731 days |                  | index = lfa height = height + 0.7 cm |
| hfa or lfa | supine       | \< 731 days |                  | index = lfa                          |
| hfa or lfa | unknown      | \< 731 days |                  | index = lfa                          |
| hfa or lfa | standing     | ≥ 731 days  |                  | index = hfa                          |
| hfa or lfa | supine       | ≥ 731 days  |                  | index = hfa height = height - 0.7 cm |
| hfa or lfa | unknown      | ≥ 731 days  |                  | index = hfa                          |
| wfh or wfl | standing     |             | \< 65 cm         | index = wfl height = height + 0.7 cm |
| wfh or wfl | standing     |             | ≥ 65 cm          | index = wfh                          |
| wfh or wfl | supine       |             | ≤ 110 cm         | index = wfl                          |
| wfh or wfl | supine       |             | more than 110 cm | index = wfh height = height - 0.7 cm |
| wfh or wfl | unknown      |             | \< 87 cm         | index = wfl                          |
| wfh or wfl | unknown      |             | ≥ 87 cm          | index = wfh                          |
| bfa        | standing     | \< 731 days |                  | height = height + 0.7 cm             |
| bfa        | standing     | ≥ 731 days  |                  | height = height - 0.7 cm             |

 

The `addWGSR()` function will not produce error messages unless there is
something very wrong with the data or the specified parameters. If an
error is encountered in a record then the value **NA** is returned.
Error conditions are listed in the table below.

 

| **Error condition**                               | **Action**                                                                 |
| ------------------------------------------------- | -------------------------------------------------------------------------- |
| Missing or nonsense value in `standing` parameter | Set `standing` to `3` (unknown) and apply appropriate height or age rules. |
| Unknown `index` specified                         | Return **NA** for z-score.                                                 |
| Missing `sex`                                     | Return **NA** for z-score.                                                 |
| Missing `firstPart`                               | Return **NA** for z-score.                                                 |
| Missing `secondPart`                              | Return **NA** for z-score.                                                 |
| `sex` is not male (`1`) or female (`2`)           | Return **NA** for z-score.                                                 |
| `firstPart` is not numeric                        | Return **NA** for z-score.                                                 |
| `secondPart` is not numeric                       | Return **NA** for z-score.                                                 |
| Missing `thirdPart` when `index = "bfa"`          | Return **NA** for z-score.                                                 |
| `thirdPart` is not numeric when `index = "bfa"`   | Return **NA** for z-score.                                                 |
| `secondPart` is out of range for specified index  | Return **NA** for z-score.                                                 |

 

We can see this error behaviour using the example data:

``` r
table(is.na(svy$wfhz))
#> 
#> FALSE  TRUE 
#>   220     1
```

We can display the problem record:

``` r
svy[is.na(svy$wfhz), ]
#>   psu age sex weight height muac oedema wfhz
#> 4   1  11   9    6.5   64.9  129      2   NA
```

The problem is due to the value **9** in the `sex` column, which should
be coded **1** (for male) and **2** (for female). Z-scores are only
calculated for records with sex specified as either **1** (male) or
**2** (female). All other values, including **NA**, will return **NA**.

The `addWGSR()` function requires that data are recorded using the
required units or required codes (see `?addWGSR` to check units required
by the different function parameters).

The `addWGSR()` function will return incorrect values if the data are
not recorded using the required units. For example, this attempt to add
weight-for-age z-scores to the example data:

``` r
svy <- addWGSR(data = svy, sex = "sex", firstPart = "weight", 
               secondPart = "age", index = "wfa")
#> ================================================================================
```

will give incorrect results:

``` r
summary(svy$wfaz)
#>    Min. 1st Qu.  Median    Mean 3rd Qu.    Max.    NA's 
#>   3.450   7.692   9.840   9.684  11.430  15.900       1
```

The odd range of values is due to age being recorded in months rather
than days.

It is simple to convert all ages from months to days:

``` r
svy$age <- svy$age * (365.25 / 12)
head(svy)
#>   psu      age sex weight height muac oedema  wfhz wfaz
#> 1   1 304.3750   1    5.7   64.2  125      2 -2.73 3.45
#> 2   1 304.3750   2    5.8   64.4  121      2 -2.04 3.95
#> 3   1 273.9375   2    6.5   62.2  139      2  0.13 5.12
#> 4   1 334.8125   9    6.5   64.9  129      2    NA   NA
#> 5   1 730.5000   2    6.5   72.9  120      2 -3.44 3.82
#> 6   1 365.2500   2    6.6   69.4  126      2 -2.26 5.01
```

before calculating and adding weight-for-age z-scores:

``` r
svy <- addWGSR(data = svy, sex = "sex", firstPart = "weight", 
               secondPart = "age", index = "wfa")
#> ================================================================================
head(svy)
#>   psu      age sex weight height muac oedema  wfhz  wfaz
#> 1   1 304.3750   1    5.7   64.2  125      2 -2.73 -4.13
#> 2   1 304.3750   2    5.8   64.4  121      2 -2.04 -3.19
#> 3   1 273.9375   2    6.5   62.2  139      2  0.13 -1.97
#> 4   1 334.8125   9    6.5   64.9  129      2    NA    NA
#> 5   1 730.5000   2    6.5   72.9  120      2 -3.44 -4.61
#> 6   1 365.2500   2    6.6   69.4  126      2 -2.26 -2.56
summary(svy$wfaz)
#>    Min. 1st Qu.  Median    Mean 3rd Qu.    Max.    NA's 
#>  -4.610  -1.873  -1.085  -1.154  -0.480   2.600       1
```

The muac column in the example dataset is recorded in millimetres (mm).
We need to convert this to centimetres (cm):

``` r
svy$muac <- svy$muac / 10
head(svy)
#>   psu      age sex weight height muac oedema  wfhz  wfaz
#> 1   1 304.3750   1    5.7   64.2 12.5      2 -2.73 -4.13
#> 2   1 304.3750   2    5.8   64.4 12.1      2 -2.04 -3.19
#> 3   1 273.9375   2    6.5   62.2 13.9      2  0.13 -1.97
#> 4   1 334.8125   9    6.5   64.9 12.9      2    NA    NA
#> 5   1 730.5000   2    6.5   72.9 12.0      2 -3.44 -4.61
#> 6   1 365.2500   2    6.6   69.4 12.6      2 -2.26 -2.56
```

before using the `addWGS()` function to calculate MUAC-for-age z-scores:

``` r
svy <- addWGSR(svy, sex = "sex", firstPart = "muac",    
               secondPart = "age", index = "mfa")
#> ================================================================================
head(svy)
#>   psu      age sex weight height muac oedema  wfhz  wfaz  mfaz
#> 1   1 304.3750   1    5.7   64.2 12.5      2 -2.73 -4.13 -1.97
#> 2   1 304.3750   2    5.8   64.4 12.1      2 -2.04 -3.19 -1.88
#> 3   1 273.9375   2    6.5   62.2 13.9      2  0.13 -1.97 -0.14
#> 4   1 334.8125   9    6.5   64.9 12.9      2    NA    NA    NA
#> 5   1 730.5000   2    6.5   72.9 12.0      2 -3.44 -4.61 -2.70
#> 6   1 365.2500   2    6.6   69.4 12.6      2 -2.26 -2.56 -1.46
```

As a last example we will use the `addWGSR()` function to add body mass
index-for-age (bfa) z-scores to the data to create a new variable called
bmiAgeZ with a precision of 4 decimal places as:

``` r
svy <- addWGSR(data = svy, sex = "sex", firstPart = "weight", 
               secondPart = "height", thirdPart = "age", index = "bfa", 
               output = "bmiAgeZ", digits = 4)
#> ================================================================================
head(svy)
#>   psu      age sex weight height muac oedema  wfhz  wfaz  mfaz bmiAgeZ
#> 1   1 304.3750   1    5.7   64.2 12.5      2 -2.73 -4.13 -1.97 -2.6928
#> 2   1 304.3750   2    5.8   64.4 12.1      2 -2.04 -3.19 -1.88 -2.0005
#> 3   1 273.9375   2    6.5   62.2 13.9      2  0.13 -1.97 -0.14  0.0405
#> 4   1 334.8125   9    6.5   64.9 12.9      2    NA    NA    NA      NA
#> 5   1 730.5000   2    6.5   72.9 12.0      2 -3.44 -4.61 -2.70 -2.8958
#> 6   1 365.2500   2    6.6   69.4 12.6      2 -2.26 -2.56 -1.46 -2.0796
```

## Usage - legacy functions

To maintain support for earlier versions of the package, the earlier
functions used to calculate anthropometric z-scores for
`weight-for-age`, `height-for-age` and `weight-for-height` still work
but will be removed (defunct) in the next version of `zscorer`. For
current and new users, it is recommended to use `addWGSR()` and
`getWGSR()` functions instead.

### Calculating z-score for each of the three anthropometric indices for a single child

For this example, we will use the `getWGS()` function and apply it to
dummy data of a **52 month** old male child with a weight of **14.6 kg**
and a height of **98.0 cm**.

``` r
# weight-for-age z-score
waz <- getWGS(sexObserved = 1,     # 1 = Male / 2 = Female
              firstPart = 14.6,    # Weight in kilograms up to 1 decimal place
              secondPart = 52,     # Age in whole months
              index = "wfa")       # Anthropometric index (weight-for-age)

waz
#> [1] -1.187651

# height-for-age z-score
haz <- getWGS(sexObserved = 1,
              firstPart = 98,      # Height in centimetres
              secondPart = 52,
              index = "hfa")       # Anthropometric index (height-for-age)

haz
#> [1] -1.741175

# weight-for-height z-score
whz <- getWGS(sexObserved = 1,
              firstPart = 14.6,
              secondPart = 98,
              index = "wfh")       # Anthropometric index (weight-for-height)

whz
#> [1] -0.1790878
```

Applying the `getWGS()` function results in a calculated `z-score` for
one child.

### Calculating z-score for each of the three anthropometric indices for a cohort or sample of children

For this example, we will use the `getCohortWGS()` function and apply it
to sample data `anthro1` that came with `zscorer`.

``` r
# Make a call for the anthro1 dataset
anthro1
```

As you will see, this dataset has the 4 variables you will need to use
with `getCohortWGS()` to calculate the `z-score` for the corresponding
anthropometric index. These are `age`, `sex`, `weight` and `height`.

``` r
head(anthro1)
#>   psu age sex weight height muac oedema   haz   waz   whz flag
#> 1   1   6   1    7.3   65.0  146      2 -1.23 -0.76  0.06    0
#> 2   1  42   2   12.5   89.5  156      2 -2.35 -1.39 -0.02    0
#> 3   1  23   1   10.6   78.1  149      2 -2.95 -1.06  0.57    0
#> 4   1  18   1   12.8   81.5  160      2 -0.28  1.42  2.06    0
#> 5   1  52   1   12.1   87.3  152      2 -4.21 -2.68 -0.14    0
#> 6   1  36   2   16.9   93.0  190      2 -0.54  1.49  2.49    0
```

To calculate the three anthropometric indices for all the children in
the sample, we execute the following commands in R:

``` r
# weight-for-age z-score
waz <- getCohortWGS(data = anthro1,
                    sexObserved = "sex",
                    firstPart = "weight",
                    secondPart = "age",
                    index = "wfa")
head(waz, 50)
#>  [1] -0.75605549 -1.39021503 -1.05597853  1.41575096 -2.67757242  1.49238050
#>  [7] -0.12987704 -0.02348159 -1.50647344 -1.54381630 -2.87495712 -0.43497240
#> [13] -1.03899540 -1.69281855 -1.31245898 -2.21003260 -0.01189226 -0.90917762
#> [19] -0.67839855 -0.94746695 -2.49960425 -0.95659644 -1.65442686 -1.25052760
#> [25]  0.67335751  0.30156301  0.24261346 -2.78670709 -1.15820651 -1.15477183
#> [31] -1.35540820 -0.59134959 -4.14967218 -0.45748752 -0.74331669 -1.69725836
#> [37] -1.05745067 -0.18869508 -0.42095770 -2.21030414 -1.30536715 -3.63778143
#> [43] -0.60662526 -0.54360470 -1.59171780 -1.74745738 -0.34803338  0.69896149
#> [49] -0.74467130  0.18924572

# height-for-age z-score
haz <- getCohortWGS(data = anthro1,
                    sexObserved = "sex",
                    firstPart = "height",
                    secondPart = "age",
                    index = "hfa")
head(haz, 50)
#>  [1] -1.2258169 -2.3475886 -2.9518041 -0.2812852 -4.2056663 -0.5387678
#>  [7] -2.4020719 -1.0317699 -2.7410884 -4.7037571 -2.5670550 -2.1144960
#> [13] -2.2323505 -2.3155458 -2.7516165 -2.7930694  0.1121349 -1.9001797
#> [19] -2.9543730 -1.9671042 -3.8716522  0.8667206 -2.8252069 -2.1412285
#> [25] -2.7994643  0.5496459 -1.4372002 -3.7979410 -2.5661752 -1.8301183
#> [31] -1.6548589 -2.7110333 -3.6399642 -1.7955069 -1.6775100 -1.0317699
#> [37] -0.4356881 -1.2660152  0.4990326 -4.6085660 -3.1662351 -1.0695930
#> [43] -1.8477936 -2.5502314 -1.8301183 -2.2755493 -3.2816532  0.4876774
#> [49] -2.4396410 -0.4794744

# weight-for-height z-score
whz <- getCohortWGS(data = anthro1,
                    sexObserved = "sex",
                    firstPart = "weight",

NEWS.md
## zscorer 0.3.1.9000

Fifth release of `zscorer`. This is a GitHub-only development release. In this
release:

### New features

* added 2000 CDC growth charts reference table

* added 2000 CDC functions to determine weight-for-age, weight-for-height,
weight-for-length, height-for-age, length-for-age, head circumference-for-age,
and BMI-for-age



## zscorer 0.3.1

Fourth release of `zscorer`. In this release:

* fixed an issue with wrong URLs in README documentation.

* fixed an issue when calculating BMI-for-age where measurements are 0 resulting in error rather than providing an NA value. Now, when values for measurements are 0, the zscore produced is NA.

* updated MUAC for age to allow for calculation for children up to 19 years old based on article by Jay Berkeley and colleagues.

* updated Shiny app and added a vignette showing how to use the Shiny app

* updated documentation to reflect editions and additions

## zscorer 0.3.0

Third release of `zscorer`. In this release:

* fixed a bug when using `zscorer` functions `addWGSR()` and `getWGSR()` inside another package in which `wgsrData` referred to by both functions and which is included as an external dataset is not lazy loaded. `wgsrData` is now added as an internal dataset

* fixed some documentation formatting.

* removed old hex sticker and added new hex sticker

## zscorer 0.2.0

Second release of `zscorer`. In this release:

* created overall function to calculate all anthropometric indices included in the WHO Growth Standards;

* updated README to illustrate use of main overall function for calculating anthropometric indices;

* added vignettes to describe calculation of anthropometric indices

* update Shiny application


## zscorer 0.1.0

Initial release of `zscorer`


R/addWGSR.R
################################################################################
#
#'
#' Calculate WHO Growth Reference z-score for given anthropometric
#' measurement/s.
#'
#' The first function, [getWGSR()], is usually called by the [addWGSR()]
#' function but could be used as a stand-alone calculator for getting z-score
#' for a given anthropometric measurement.
#'
#' [addWGSR()] adds the WHO Growth Reference z-scores to a data frame of
#' anthropometric data for weight, height or length, MUAC, head circumference,
#' sub-scapular skinfold, triceps skinfold, and body mass index (BMI).
#'
#' @param data A survey dataset as a data.frame object
#' @param sex Name of variable specifying the sex of the subject. This must be
#'   coded as `1 = male` and `2 = female`. Give a quoted variable
#'   name as in (e.g.) `"sex"`.
#'
#' @param firstPart Name of variable specifying:
#'   * Weight (kg) for BMI/A, W/A, W/H, or W/L
#'   * Head circumference (cm) for HC/A
#'   * Height (cm) for BMI/A for H/A
#'   * Length (cm) for L/A
#'   * MUAC (cm) for MUAC/A
#'   * Sub-scapular skinfold (mm) for SSF/A
#'   * Triceps skinfold (mm) for TSF/A
#'
#'   Give a quoted variable name as in (e.g.) `"weight"`. Be careful with
#'   units (weight in kg; height, length, head circumference, and MUAC in cm,
#'   skinfolds in mm).
#'
#' @param secondPart Name of variable specifying:
#'   * Age (days) for H/A, HC/A, L/A, MUAC/A, SSF/A, or TSF/A
#'   * Height (cm) BMI/A or W/H
#'   * Length (cm) for W/L
#'
#'   Give a quoted variable name as in (e.g.) `"age"`. Be careful with
#'   units (age in days; height and length in cm).
#'
#' @param thirdPart Name of variable specifying age (in days) for BMI/A. Give a
#'   quoted variable name as in (e.g.) `"age"`. Be careful with units
#'   (age in days).
#'
#' @param index The index to be calculated and added to `data`. One of:
#'
#'   | **Index** | **Definition** |
#'   | :--- | :--- |
#'   | `bfa` | BMI for age |
#'   | `hca` | Head circumference for age |
#'   | `hfa` | Height for age |
#'   | `lfa` | Length for age |
#'   | `mfa` | MUAC for age |
#'   | `ssa` | Sub-scapular skinfold for age |
#'   | `tsa` | Triceps skinfold for age |
#'   | `wfa` | Weight for age |
#'   | `wfh` | Weight for height |
#'   | `wfl` | Weight for length |
#'
#'  Give a quoted index name as in (e.g.) `"wfh"`.
#'
#' @param standing Variable specifying how stature was measured. If NULL then
#'   age (for `"hfa"` or `"lfa"`) or height rules (for `"wfh"` or `"wfl"`) will
#'   be applied. This must be coded as `1 = Standing`; `2 = Supine`;
#'   `3 = Unknown`. All other values will be recoded to `3 = Unknown`. Give a
#'   quoted variable name as in (e.g.) `"measured"` or a single value (e.g.
#'   `"measured = 1`). If no value (or NULL) is specified then height and age
#'   rules will be applied.
#'
#' @param output The name of the column containing the specified index to be
#'   added to the dataset. This is an optional parameter. If you do not specify
#'   a value for output then the added column will take the name of the
#'   specified index with a `"z"` appended.
#'
#' @param digits The number of decimal places for `output`. Defaults to 2 d.p.
#'
#' @return A data.frame of the survey dataset with the calculated z-scores
#'   added.
#'
#' @examples
#' # Given a male child 10 months old with a weight of 5.7 kgs, height of 64.2
#' # cms, and MUAC of 125 mm:
#' #
#' # Calculate weight-for-height
#' getWGSR(sex = 1,
#'         firstPart = 5.7,
#'         secondPart = 64.2,
#'         index = "wfh",
#'         standing = 3)
#'
#' # calculate weight-for-age
#' getWGSR(sex = 1,
#'         firstPart = 5.7,
#'         secondPart = 10,
#'         index = "wfa",
#'         standing = 3)
#'
#' # calculate height-for-age
#' getWGSR(sex = 1,
#'         firstPart = 64.2,
#'         secondPart = 10,
#'         index = "hfa",
#'         standing = 3)
#'
#' # Calculate MUAC-for-age z-score for a girl
#' getWGSR(sex = 1,
#'         firstPart = 20,
#'         secondPart = 62 * (365.25 / 12),
#'         index = "mfa")
#'
#' @rdname wgs
#' @export
#'
#
################################################################################

getWGSR <- function(sex,
                    firstPart, secondPart, thirdPart = NA,
                    index = NA, standing = NA) {
  ## Get WGS reference data
  wgsrData <- referenceData[["wgs"]]
  ## Avoid missing and impossible values in 'standing' by coding NA and
  ## other values to '3'
  if(is.na(standing) | !(standing %in% c(1, 2, 3))) {
    standing = 3
  }
  ## Unknown index specified - return NA
  if(!(index %in% c("bfa", "hca", "hfa", "lfa", "mfa",
                    "ssa", "tsa", "wfa", "wfh", "wfl"))) {
    return(NA)
  }
  ## Missing data for 'sex', 'firstPart', or 'secondPart' - return NA
  if(is.na(sex) | is.na(firstPart) | is.na(secondPart)) {
    return(NA)
  }
  ## 'sex' must be male (1) or female (2)
  if(!(sex %in% c(1, 2))) {
    return(NA)
  }
  ## 'firstPart' or 'secondPart' are not numeric - return NA
  if(!is.numeric(firstPart) | !is.numeric(secondPart)) {
    return(NA)
  }
  ## Missing 'thirdPart' (age) is missing for BMI-for-age - return NA
  if(index == "bfa" & is.na(thirdPart)) {
    return(NA)
  }
  ## 'thirdPart' (age) is not numeric for BMI-for-age - return NA
  if(index == "bfa" & !is.numeric(thirdPart)) {
    return(NA)
  }
  ## 'secondPart' is zero then BMI cannot be calculated
  if(index == "bfa" & secondPart == 0) {
    return(NA)
  }
  ## Round lengths to nearest 0.1 cm
  if(index %in% c("wfh", "wfl")) {
    secondPart <- round(secondPart, 1)
  }
  ## Round ages to the nearest day
  if(index %in% c("hca", "hfa", "lfa", "mfa", "ssa", "tsa", "wfa")) {
    secondPart <- round(secondPart, digits = 0)
  }
  if(index == "bfa") {
    thirdPart <- round(thirdPart, 0)
  }
  ## Rules for length-for-age and height-for-age indices
  if(standing == 1 & (index == "lfa" | index == "hfa") & secondPart < 731) {
    index <- "lfa"
    firstPart <- firstPart + 0.7
  }
  if(standing == 2 & (index == "lfa" | index == "hfa") & secondPart < 731) {
    index <- "lfa"
  }
  if(standing == 3 & (index == "lfa" | index == "hfa") & secondPart < 731) {
    index <- "lfa"
  }
  if(standing == 1 & (index == "lfa" | index == "hfa") & secondPart >= 731) {
    index <- "hfa"
  }
  if(standing == 2 & (index == "lfa" | index == "hfa") & secondPart >= 731) {
    index <- "hfa"
    firstPart <- firstPart - 0.7
  }
  if(standing == 3 & (index == "lfa" | index == "hfa") & secondPart >= 731) {
    index <- "hfa"
  }
  ## Rules for weight-for-length and weight-for-height indices
  if(standing == 1 & (index == "wfl" | index == "wfh") & secondPart < 65) {
    index = "wfl"
    secondPart <- secondPart + 0.7
  }
  if(standing == 1 & (index == "wfl" | index == "wfh") & secondPart >= 65) {
    index = "wfh"
  }
  if(standing == 2 & (index == "wfl" | index == "wfh") & secondPart <= 110) {
    index = "wfl"
  }
  if(standing == 2 & (index == "wfl" | index == "wfh") & secondPart > 110) {
    index = "wfh"
    secondPart <- secondPart - 0.7
  }
  if(standing == 3 & (index == "wfl" | index == "wfh") & secondPart < 87) {
    index = "wfl"
  }
  if(standing == 3 & (index == "wfl" | index == "wfh") & secondPart >= 87) {
    index = "wfh"
  }
  ## Rules for BMI-for-age index
  if(standing == 1 & index == "bfa" & thirdPart < 731) {
    secondPart <- secondPart + 0.7
  }
  if(standing == 2 & index == "bfa" & thirdPart >= 731) {
    secondPart <- secondPart - 0.7
  }
  ## Calculate BMI (as 'firstPart') and place age in 'secondPart'
  if(index == "bfa") {
    firstPart <- firstPart / (secondPart / 100)^2
    secondPart <- thirdPart
  }
  ## 'secondPart' is out of range for specified 'index' - return NA
  rangeSecondPart <- range(wgsrData$given[wgsrData$index == index])
  if(secondPart < rangeSecondPart[1] | secondPart > rangeSecondPart[2]) {
    return(NA)
  }
  ## Lookup reference values and calculate z-score
  lkpIndexSex <- wgsrData[wgsrData$index == index & wgsrData$sex == sex, ]
  L <- approx(lkpIndexSex$given, lkpIndexSex$l,
              xout = secondPart, ties = "ordered")$y
  M <- approx(lkpIndexSex$given, lkpIndexSex$m,
              xout = secondPart, ties = "ordered")$y
  S <- approx(lkpIndexSex$given, lkpIndexSex$s,
              xout = secondPart, ties = "ordered")$y
  z <- (((firstPart / M) ^ L) - 1) / (L * S)
  SD3pos <- M * (1 + L * S * (+3))^(1 / L)
  SD2pos <- M * (1 + L * S * (+2))^(1 / L)
  SD23pos <- SD3pos - SD2pos
  SD3neg <- M * (1 + L * S * (-3))^(1 / L)
  SD2neg <- M * (1 + L * S * (-2))^(1 / L)
  SD23neg <- SD2neg - SD3neg
  if(z >  3) z <-  3 + ((firstPart - SD3pos) / SD23pos)
  if(z < -3) z <- -3 + ((firstPart - SD3neg) / SD23neg)
  return(z)
}


################################################################################
#
#'
#' @examples
#' # Calculate weight-for-height (wfh) for the anthro3 dataset
#' addWGSR(data = anthro3,
#'         sex = "sex",
#'         firstPart = "weight",
#'         secondPart = "height",
#'         index = "wfh")
#'
#' # Calculate weight-for-age (wfa) for the anthro3 dataset
#' addWGSR(data = anthro3,
#'         sex = "sex",
#'         firstPart = "weight",
#'         secondPart = "age",
#'         index = "wfa")
#'
#' # Calculate height-for-age (hfa) for the anthro3 dataset
#' addWGSR(data = anthro3,
#'         sex = "sex",
#'         firstPart = "height",
#'         secondPart = "age",
#'         index = "hfa")
#'
#' # Calculate MUAC-for-age (mfa) for the anthro4 dataset
#'
#' ## Convert age in anthro4 from months to days
#' testData <- anthro4
#' testData$age <- testData$agemons * (365.25 / 12)
#'
#' addWGSR(data = testData,
#'         sex = "sex",
#'         firstPart = "muac",
#'         secondPart = "age",
#'         index = "mfa")
#'
#' @rdname wgs
#' @export
#'
#
################################################################################

addWGSR <- function(data, sex,
                    firstPart, secondPart, thirdPart = NA,
                    index = NA, standing = NULL,
                    output = paste(index, "z", sep = ""), digits = 2) {
  ## If 'standing' is not specified then create a column in 'data' holding
  ## 3 (unknown) for all rows
  addedStanding <- FALSE
  if(is.null(standing)) {
    ## Random column name for 'standing'
    standing <- paste(sample(c(letters, LETTERS), size = 16, replace = TRUE),
                      collapse = "")
    data[[standing]] <- 3
    addedStanding <- TRUE
  }
  ## Calculate specified index
  z <- vector(mode = "numeric", length = nrow(data))
  pb <- txtProgressBar(min = 0, max = nrow(data), style = 1)
  for(i in seq_len(nrow(data))) {
    z[i] <- ifelse(!is.na(thirdPart),
                   getWGSR(sex = data[[sex]][i],
                           firstPart = data[[firstPart]][i],
                           secondPart = data[[secondPart]][i],
                           index = index, standing = data[[standing]][i],
                           thirdPart = data[[thirdPart]][i]),
                   getWGSR(sex = data[[sex]][i],
                           firstPart = data[[firstPart]][i],
                           secondPart = data[[secondPart]][i],
                           index = index, standing = data[[standing]][i]))
    setTxtProgressBar(pb, i)
  }
  cat("\n", sep = "")
  ## Remove	added 'standing' column from 'data'
  if(addedStanding) {
    data[[standing]] <- NULL
  }
  ## Add index to 'data' and return 'data'
  data[[output]] <- round(z, digits = digits)
  return(data)
}


